src/ccache/util/file.cpp

```diff
@@ -100,9 +104,25 @@ copy_file(const fs::path& src, const fs::path& dest, ViaTmpFile via_tmp_file)
         FMT("Failed to open {} for writing: {}", dest, strerror(errno)));
     }
   }
+
+#if defined(__linux__)
+  DirEntry dir_entry(src, *src_fd);
+  if (!dir_entry) {
+    return tl::unexpected(FMT("Failed to stat: {}", src));
+  }
+  ssize_t bytes_left = dir_entry.size();
+  while (bytes_left > 0) {
+    ssize_t n = sendfile(*dest_fd, *src_fd, nullptr, bytes_left);
+    if (n < 0) {
+      return tl::unexpected(FMT("Failed to copy: {} to {}", src, dest));
+    }
```

Comment:
The [sendfile man page](https://man7.org/linux/man-pages/man2/sendfile.2.html) says:

> Applications may wish to fall back to read(2) and write(2) in the case where sendfile() fails with EINVAL or ENOSYS.

Do you think that would make sense?